### PR TITLE
Add autograd API to at::Tensor

### PR DESCRIPTION
**Pull request by goldsborough**
This PR adds a basic autograd API to `at::Tensor`, and implements it for `Variable`. This API consists of:

- `void set_requires_grad(bool)`,
- `bool requires_grad() const`
- `Tensor& grad()`
- `const Tensor& grad() const`
- `Tensor detach() const`
- `void detach_()`

It does this in three steps:

1. Adds these functions as virtual methods to `TensorImpl`, and makes them throw exceptions when called,
2. Implements these functions in `VariableImpl`, mostly moving code from `Variable` into `VariableImpl`
3. Adds the methods to `Tensor` such that `Tensor::requires_grad` calls `pImpl->requires_grad()`
4. Solves all the problems that arise along the way.

One extra thing: In Python, `module.cuda()` works by changing the data and type underneath `Variable`'s feet (i.e. changing the underlying tensor data), so that references to `Variable` are not invalidated. The same is required in C++, so as part of this autograd API I have implemented `at::detail::set_data(Tensor& tensor, Tensor new_data)`, which assigns `new_data` to `tensor.data()` (`tensor` is here actually a `Variable`) and changes the type of the variable (`tensor`). To avoid `set_data` being part of the public API of `at::Tensor`, I made it a free friend function in the  `at::detail` namespace. That should hide it well. See the change in `torch/csrc/autograd/python_variable.cpp` for how this simplifies the code.

Because of name collisions in `VariableImpl`, e.g. between the `requires_grad` function and the `requires_grad` member variable, I had to rename some members (4 or 5) of `VariableImpl` to give them trailing underscores. For consistency, I then added trailing underscores to all member variables. This creates a little noise in this PR -- I'm sorry. I can revert those changes for now if it's too annoying to review.

@colesbury @apaszke @ezyang

### CODE REVIEW

File: aten/src/ATen/templates/Tensor.h [hunk 1]
```diff
@@ -120,6 +126,33 @@ struct Tensor : public detail::TensorBase {
...
+  Tensor detach() const {
+    return pImpl->detach();
+  }
+  void detach_() {
+    pImpl->detach_();
+  }
+
+  friend void detail::set_data(Tensor& tensor, Tensor new_data) {
+    tensor.pImpl->set_data(new_data);
+  }
```
Comment (apaszke) [inline aten/src/ATen/templates/Tensor.h]: Wow I had no idea you can declare friend functions in the body of a struct

File: torch/csrc/autograd/functions/accumulate_grad.cpp [hunk 2]
```diff
@@ -30,10 +30,11 @@ auto AccumulateGrad::apply(const variable_list& grads) -> variable_list {
     new_grad = (*hook)({new_grad})[0];
   }
 
-  auto& grad = variable.grad();
-  if (!grad.defined()) {
+  auto& grad_tensor = variable.grad();
```
Comment (apaszke) [inline torch/csrc/autograd/functions/accumulate_grad.cpp]: Can we please keep it's name as `grad` and just make the type `at::Tensor`? We usually call things `_tensor` when they are not variables (because the type system doesn't help us), and this will be confusing

File: torch/csrc/autograd/variable.h [hunk 3]
```diff
@@ -333,40 +303,82 @@ struct Variable::Impl : public at::TensorImpl {
 
   std::shared_ptr<Function> get_grad_accumulator();
   virtual std::shared_ptr<Function>& get_grad_fn() {
-    return grad_fn;
+    return grad_fn_;
+  }
+
+  virtual const Variable& base() const {
+    throw std::runtime_error("Can't get base of non-view Variable");
+  }
```
Comment (apaszke) [inline torch/csrc/autograd/variable.h]: Why do we need this in `VariableImpl` now?

File: torch/csrc/autograd/variable.h [hunk 4]
```diff
@@ -333,40 +303,82 @@ struct Variable::Impl : public at::TensorImpl {
...
+
+  /// Sets the `requires_grad` property of `Variable`. This should be true for
+  /// leaf variables that want to accumulate gradients, and false for all other
+  /// variables.
+  void set_requires_grad(bool requires_grad) noexcept override {
```
Comment (apaszke) [inline torch/csrc/autograd/variable.h]: Can you please mark inherited virtual functions as virtual?

**Review by apaszke — COMMENTED** [review]
A bunch of things that were non-virtual previously became virtual now (most notably `requires_grad()`). Are there any perf implications? Have you ran any benchmarks?

File: torch/csrc/autograd/variable.h [hunk 5]
```diff
@@ -333,40 +303,82 @@ struct Variable::Impl : public at::TensorImpl {
...
+    requires_grad_ = requires_grad;
+  }
+
+  bool requires_grad() const noexcept override {
+    return requires_grad_ || grad_fn_ || (is_view_ && base().requires_grad());
```
Comment (apaszke) [inline torch/csrc/autograd/variable.h]: This is virtual now, so we don't need `is_view_` anymore. It was there only to be a faster alternative to virtual dispatch.

File: aten/src/ATen/templates/Tensor.h [hunk 6]
```diff
@@ -120,6 +126,33 @@ struct Tensor : public detail::TensorBase {
...
+  // ~~~~~ Autograd API ~~~~~
+
+  void set_requires_grad(bool requires_grad) {
+    pImpl->set_requires_grad(requires_grad);
+  }
+  bool requires_grad() const noexcept {
+    return pImpl->requires_grad();
+  }
+
+  Tensor& grad() noexcept {
```
Comment (ezyang) [inline aten/src/ATen/templates/Tensor.h]: I would much prefer a `set_grad()` virtual method, as opposed to a method that returns a reference that can be modified.

Comment (ezyang) [reply aten/src/ATen/templates/Tensor.h]: Also, all of this const business isn't really getting anything useful done, c.f. https://github.com/zdevito/ATen/issues/27

**Comment by ezyang** [comment]
You don't have to do it, but what's stopping us from getting rid of `Variable` altogether after this patch?

File: torch/csrc/autograd/variable.h (hunk 4 above)
Comment (apaszke) [reply torch/csrc/autograd/variable.h]: I know it's redundant but it's incredibly important. I never understand why people don't propagate this keyword, because it completely changes the semantics and perf characteristic of a function.

Comment (apaszke) [reply torch/csrc/autograd/variable.h]: Ohh I see now. Never mind then, that's good!

**Comment by apaszke** [comment]
@ezyang We have a tonne of accessors and functions for graph management in Variables, so I don't think we're ready to remove them just yet.

File: torch/csrc/autograd/variable.h (hunk 5 above)
Comment (goldsborough) [reply torch/csrc/autograd/variable.h]: Yeah but right now `base()` will throw an exception when called for a non-view. And if I change  `base()` to return `*this` instead, then `base().requires_grad()` will infinite loop. So unless I do a `dynamic_cast`, I'm not sure there's a better way to solve it right now. What do you think?

File: torch/csrc/autograd/variable.h (hunk 3 above)
Comment (goldsborough) [reply torch/csrc/autograd/variable.h]: I need to call `base()` from `requires_grad()`

File: aten/src/ATen/templates/Tensor.h (hunk 6 above)
Comment (goldsborough) [reply aten/src/ATen/templates/Tensor.h]: I think it's still useful once you downcast the `Tensor&` to `Variable&`, since `const` is useful for `Variable`

**Comment by apaszke** [comment]
Have you ran any benchmarks? I'm a little concerned about getting this in before 0.4 is released because of possible perf implications.

**Comment by goldsborough** [comment]
No I haven't because of Ed's comment. Do you want to microbenchmark the methods or benchmark WLM or something?

File: torch/csrc/autograd/variable.h (hunk 3 above)
Comment (apaszke) [reply torch/csrc/autograd/variable.h]: Can't we access the pointer directly?

File: torch/csrc/autograd/variable.h (hunk 5 above)
Comment (apaszke) [reply torch/csrc/autograd/variable.h]: Ugh... Ok, let's keep it.

**Comment by apaszke** [comment]
I think both will be useful. I would like to avoid us harming the perf right before the release, without any time to fix it.

**Comment by goldsborough** [comment]
Isn't this what @yf225's perf tests do? Can we just look at what they say at the end of the tests? I will write some function-level benchmarks in the meantime

**Comment by apaszke** [comment]
I don't know if they e.g. run WLM with cuDNN disabled (it becomes a single large op otherwise). Some C++ microbenchmarks would be useful too, and we don't have those except some old code in the `benchmark` repo.

**Comment by yf225** [comment]
Currently in the GPU perf tests, `cudnn_lstm`, `word_language_model` and `mnist` are using cuDNN, while `lstm` and `mlstm` are not

**Comment by goldsborough** [comment]
Microbenchmarks for functions. Time in microseconds. I see an average slowdown of 0.001-0.006 microseconds = 1-6 nanoseconds. @apaszke I think it's ok?

```
Before:

requires_grad: average time: 0.108163 | min time: 0.105 | max time: 0.441

set_requires_grad: average time: 0.11338 | min time: 0.102 | max time: 7.73

grad: average time: 0.146497 | min time: 0.139 | max time: 0.281

After:

requires_grad: average time: 0.114218 | min time: 0.109 | max time: 0.28

set_requires_grad: average time: 0.119656 | min time: 0.105 | max time: 11.184

grad: average time: 0.147445 | min time: 0.14 | max time: 0.354
```

File: aten/src/ATen/TensorImpl.h [hunk 7]
```diff
@@ -50,9 +51,30 @@ struct TensorImpl : public Retainable {
   void setScalar(bool s) {
     is_scalar = s;
   }
+
+  // ~~~~~ Autograd API ~~~~~
+  // Some methods below are defined in TensorImpl.cpp because Tensor is an
+  // incomplete type.
```
Comment (apaszke) [inline aten/src/ATen/TensorImpl.h]: Looks like they're all methods of `Tensor` too

File: torch/csrc/autograd/functions/accumulate_grad.cpp [hunk 8]
```diff
@@ -30,19 +30,20 @@ auto AccumulateGrad::apply(const variable_list& grads) -> variable_list {
     new_grad = (*hook)({new_grad})[0];
   }
 
-  auto& grad = variable.grad();
+  at::Tensor& grad = variable.grad();
   if (!grad.defined()) {
     variable.grad() = new_grad.clone();
   } else if (!GradMode::is_enabled()) {
+    Variable& grad_variable = as_variable_ref(grad);
```
Comment (apaszke) [inline torch/csrc/autograd/functions/accumulate_grad.cpp]: Is there any reason why we couldn't just `as_variable_ref` when it's defined?

**Review by apaszke — COMMENTED** [review]
LGTM. cc @colesbury who might have some opinions too

**Review by apaszke — APPROVED** [review]
(Was meant to be an approve)

**Comment by apaszke** [comment]
Can you only slightly change the code so that it takes an `at::Tensor` input and benchmarks on this? Then you could create a variable in a different compilation unit and pass it in. Just to make sure it didn't get devirtualized because the numbers are extremely close.

**Comment by goldsborough** [comment]
Ok will try 👍

**Comment by ezyang** [comment]
Re Windows error you're probably missing an `AT_API`.

File: aten/src/ATen/TensorImpl.h [hunk 9]
```diff
@@ -50,9 +51,30 @@ struct TensorImpl : public Retainable {
...
+
+  AT_API virtual Tensor& grad();
+  AT_API virtual const Tensor& grad() const;
+
+  AT_API virtual Tensor detach() const;
+  AT_API virtual void detach_() {
+    AT_ERROR("detach_ is not implemented for Tensor");
+  }
+
+  virtual void set_data(Tensor new_data);
```
Comment (yf225) [inline aten/src/ATen/TensorImpl.h]: @goldsborough I think it's probably missing `AT_API` here

Comment (goldsborough) [reply aten/src/ATen/TensorImpl.h]: Thanks!

**Comment by goldsborough** [comment]
CC @ebetica @jgehring this PR means `at::Tensor` becomes the defacto Variable type

**Comment by goldsborough** [comment]
Yeah, sorry, let me rephrase this to "`at::Tensor` becomes the defacto Variable type __in the public C++ API__". As you know, PyTorch 0.4 exposes only tensors *with the ability to record gradients* (where a tensor with `requires_grad=False` is like a non-gradient recording Tensor, effectively). We want the same for C++, i.e. there should only be one *public* tensor type, which will be called `torch::Tensor`, but is really `autograd::Variable`. Now, we need the type we expose to users to be `at::Tensor` (which we call `torch::Tensor`), since the whole ATen library uses `at::Tensor`. But when users create tensors, they really create `autograd::Variable`s, which are upcast to `at::Tensor`. The way we brought back gradient-related functionality to users in the C++ extensions API is to then have free functions, e.g. `void set_requires_grad(at::Tensor)`, which inside would downcast the tensor to a Variable and call `.set_requires_grad()` on the downcast tensor. This is because we didn't want users to know of the distinction between variables and tensors -- they should only see tensors (like in Python). What this PR now does is instead put the gradient-related functions as virtual methods on the `at::Tensor` type directly, so that we can call `set_requires_grad` on `at::Tensor`, which will work when the `at::Tensor` is actually a variable, which it always should be for public users, and throw an exception for the base tensor, since it remains non-gradient-recording. Does this make sense?

**Comment by ebetica** [comment]
Cool! So we still make Variables whenever we want them to record gradients like usual then? So make_variable should just return a Tensor?

**Comment by goldsborough** [comment]
Not directly, instead we created `torch::CPU` and `torch::CUDA` which return `VariableType`s such that e.g. `at::ones(torch::CPU(kFloat), {5})` gives you a variable (upcast to Tensor) -- see https://github.com/pytorch/pytorch/blob/master/test/cpp_extensions/extension.cpp#L9
